### PR TITLE
Introducing Upload Strategies 

### LINKED ISSUE #760: Create new resources on the server using POST
We need to support [create](https://hl7.org/fhir/http.html#create) using POST in the sync API.

<s>
**Describe the bug**
Locally created Resources are assigned ID at creation/before saving to the local db. Once this Resource is up-synced, the server assigns a new ID to the resource and disregards the assigned ID  (in case of HTTP-POST). The created object is returned  in the POST response. 
Since the lD for local & remote copy are different now, need to update all the local id references with the remote id.

**Component**
Core library, SDC library, reference app, or SDC gallery app

**To Reproduce**

**Expected behavior**

**Screenshots**

**Smartphone (please complete the following information):**
 - Device: [e.g. Pixel4a emulator]
 - Android version: [e.g. Settings -> About phone -> Android version]
 - Build number: [e.g. Settings -> About phone -> Build number]

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.
</s>

### PR DESCRIPTION
**Pull request by anchita-g**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

One of the PRs to fix #760 

**Description**
- Created a new class `PatchGeneratorMode` to define the different modes for PatchGeneration
- Created a new class `UploadStrategy` to enlist all the possible options to support Upload 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [ ] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [ ] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [ ] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [ ] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [ ] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [ ] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * limitations under the License.
+ */
+
+package com.google.android.fhir.sync.upload
+
+import com.google.android.fhir.sync.upload.patch.PatchGeneratorMode
+import com.google.android.fhir.sync.upload.request.UploadRequestGeneratorMode
+import org.hl7.fhir.r4.model.codesystems.HttpVerb
+
+sealed class UploadStrategy(
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt:23]: We do not want to allow developers to construct their own UploadStrategy. Let's ensure that by making the constructor private
```suggestion
sealed class UploadStrategy private constructor(
```

File: engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.fhir.sync.upload
+
+import com.google.android.fhir.sync.upload.patch.PatchGeneratorMode
+import com.google.android.fhir.sync.upload.request.UploadRequestGeneratorMode
+import org.hl7.fhir.r4.model.codesystems.HttpVerb
+
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt:22]: add kdoc

**Review by omarismail94 — COMMENTED** [review]
Nice change Anchita!

File: engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt (hunk 1 above)
Comment (anchita-g) [reply engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt:23]: Makes sense, the constructor for sealed class is `protected` by default.

**Review by omarismail94 — APPROVED** [review]
Great change @anchita-g !

**Pull request merged**